🌐  Add Dutch translation for `docs/nl/docs/features.md` (#12101)
fastapi/fastapi: docs/nl/docs/features.md

![ReDoc](https://fastapi.tiangolo.com/img/index/index-06-redoc-02.png)

### Gewoon Moderne Python

Het is allemaal gebaseerd op standaard **Python type** declaraties (met dank aan Pydantic). Je hoeft dus geen nieuwe syntax te leren. Het is gewoon standaard moderne Python.

Als je een opfriscursus van 2 minuten nodig hebt over het gebruik van Python types (zelfs als je FastAPI niet gebruikt), bekijk dan een korte tutorial: [Python Types](python-types.md){.internal-link target=_blank}.

Je schrijft gewoonn standaard Python met types:

```Python
from datetime import date

from pydantic import BaseModel

# Declareer een variabele als een str
# en krijg editorondersteuning in de functie
def main(user_id: str):
    return user_id

# Een Pydantic model
class User(BaseModel):
    id: int
    name: str
    joined: date
```

Vervolgens kan je het op deze manier gebruiken:

```Python
my_user: User = User(id=3, name="John Doe", joined="2018-07-19")

second_user_data = {
    "id": 4,
    "name": "Mary",
    "joined": "2018-11-30",
}

my_second_user: User = User(**second_user_data)
```

/// info

`**second_user_data` betekend:

> Geef de keys en values van de `second_user_data` dict direct door als key-value argumenten, gelijk aan: `User(id=4, name=“Mary”, joined=“2018-11-30”)`

Reviewer: "keys", "values" & "dict" still need to be translated. Perhaps to "sleutels", "waarden" & "mapping"?

Author: I found this a difficult one. Every Dutch speaking python programmer I know talks about "keys" and "values" in the context of a dictionary. In my view, FastAPI is aimed at people that have basic understanding of python programming. Renaming them to "sleutels" and "waarden" would, in my opinion, create unnecessary confusion.

Maybe use the Dutch words but add the English transaction after it in parentheses (or the other way around)?

```markdown
Geef de sleutels (keys) en waarden  values van de `second_user_data` dict direct door als sleutel-waarden argumenten, gelijk aan: `User(id=4, name=“Mary”, joined=“2018-11-30”)`
```

Reviewer: Yea, I get what you mean. I honestly find some of these translations a bit awkward myself. But your solution sounds good to me, so something like "sleutels (keys) en waarden (values)" should work.

> Geef de sleutels (keys) en waarden (values) de `second_user_data` dict direct door als sleutel-waarden argumenten, gelijk aan: `User(id=4, name=“Mary”, joined=“2018-11-30”)`

Reviewer: ```suggestion
Geef de sleutels (keys) en waarden (values) van de `second_user_data` dict direct door als sleutel-waarden argumenten, gelijk aan: `User(id=4, name=“Mary”, joined=“2018-11-30”)`
```

///

### Editor-ondersteuning

Het gehele framework is ontworpen om eenvoudig en intuïtief te zijn in gebruik. Alle beslissingen zijn getest op meerdere code-editors nog voordat het daadwerkelijke ontwikkelen begon, om zo de beste ontwikkelervaring te garanderen.

Uit enquêtes onder Python ontwikkelaars blijkt maar al te duidelijk dat "autocompletion" <a href="https://www.jetbrains.com/research/python-developers-survey-2017/#tools-and-features" class="external-link" target="_blank">een van de meest gebruikte functionaliteiten is</a>.

Het hele **FastAPI** framework is daarop gebaseerd. Autocompletion werkt overal.

Je hoeft zelden terug te vallen op de documentatie.

Zo kan je editor je helpen:

* in <a href="https://code.visualstudio.com/" class="external-link" target="_blank">Visual Studio Code</a>:

![editor ondersteuning](https://fastapi.tiangolo.com/img/vscode-completion.png)

* in <a href="https://www.jetbrains.com/pycharm/" class="external-link" target="_blank">PyCharm</a>:

![editor ondersteuning](https://fastapi.tiangolo.com/img/pycharm-completion.png)

> Je krijgt autocomletion die je voorheen misschien zelfs voor onmogelijk had gehouden. Zoals bijvoorbeeld de `price` key in een JSON body (die genest had kunnen zijn) die afkomstig is van een request.

Reviewer: "autocompletion" has been translated to "(automatische) code aanvulling" in #12042

Je hoeft niet langer de verkeerde keys in te typen, op en neer te gaan tussen de documentatie, of heen en weer te scrollen om te checken of je `username` of toch `user_name` had gebruikt.

### Kort

Dit framework heeft voor alles verstandige **standaardinstellingen**, met overal optionele configuraties. Alle parameters kunnen worden verfijnd zodat het past bij wat je nodig hebt, om zo de API te kunnen definiëren die jij nodig hebt.

Maar standaard werkt alles **“gewoon”**.

### Validatie

* Validatie voor de meeste (of misschien wel alle?) Python **datatypes**, inclusief:
    * JSON objecten (`dict`).
    * JSON array (`list`) die itemtypes definiëren.
    * String (`str`) velden, die min en max lengtes hebben.
    * Getallen (`int`, `float`) met min en max waarden, etc.

* Validatie voor meer exotische typen, zoals:
    * URL.
    * E-mail.
    * UUID.
    * ...en anderen.

Alle validatie wordt uitgevoerd door het beproefde en robuuste **Pydantic**.

### Beveiliging en authenticatie

> Beveiliging en authenticatie is geïntegreerd. Zonder compromissen te doen naar databases of datasmodellen.

Reviewer: ```suggestion
Beveiliging en authenticatie is geïntegreerd. Zonder compromissen te doen naar databanken of datamodellen.
```

Author: Isn't database also a Dutch word? I think it's in the dictionary.

Reviewer: Maybe, Dutch has a lot of English "loan words", but I think "databank" is a very clear Dutch word that is used a lot in Dutch text books and other literature, so I would definitely prefer it over "database".

Author: I think you are right. The Dutch spoken in The Netherlands uses more English words in comparison to the Dutch spoken in for example Belgium.

 At universities in The Netherlands, Dutch spoken computer science (informatica) programs use "databases".

- https://www.uu.nl/bachelors/informatica/studieprogramma
- https://www.uva.nl/programmas/bachelors/informatica/studieprogramma/studieprogramma.html

> e.g. first site I look, informatica opleiding at UGent, has a course called "databanken": https://informatica.ugent.be/bachelor/

Yes that's indeed the case. However, "Databases" is also the name of several courses at Belgian universities.

- https://www.uantwerpen.be/nl/studeren/aanbod/alle-opleidingen/informatica-studeren/bachelor/studieprogramma/
- https://www.uhasselt.be/nl/studeren/opleidingen/bachelor-en-master-in-de-informatica#anch-215-bachelor

The most recent edition of the Dutch book ""Principes van databanken" by Guy de Tré that is used in the UGent course is published with the title "Principes van databases".

The search term for "database" is also more popular than "databank" in both Belgium and The Netherlands.

- https://trends.google.com/trends/explore?date=all&geo=BE&q=database,databank&hl=nl-NL
- https://trends.google.com/trends/explore?date=all&geo=NL&q=database,databank&hl=nl-NL

So I don't really think there is a correct definitive correct Dutch translation of "databases".

However, in my view, the term “database” is more often used in Dutch (in Belgium and The Netherlands, in Universities, in text books and in government).

I think "database" leads to less confusion for the user when also taking into account that they are probably also reading other (English) documentation.

Reviewer: > So I don't really think there is a correct definitive correct Dutch translation of "databases".

Haha - I respectfully disagree: the clear correct Dutch translation of "database" is "databank", but the question is whether the actual Dutch translation, or the English loan word, is used more frequently in Dutch texts 🤷 

Anyway, why don't we adhere to this rule of thumb in cases like these: to follow the Dutch-speaking Wikipedia? While that's usually also a bit skewed towards The Netherlands vs Belgium, at least it's an "authorative" source to use as gold truth.

In this particular case: that would mean we stick to "database" as https://nl.wikipedia.org/wiki/Databank redirects to https://nl.wikipedia.org/wiki/Database.

Please do go through my reviewer's comments carefully though - some of the "databank" suggestions also contain fixes for other words 😉

Alle beveiligingsschema's gedefinieerd in OpenAPI, inclusief:

* HTTP Basic.
* **OAuth2** (ook met **JWT tokens**). Bekijk de tutorial over [OAuth2 with JWT](tutorial/security/oauth2-jwt.md){.internal-link target=_blank}.
* API keys in:
    * Headers.
    * Query parameters.
    * Cookies, etc.

Plus alle beveiligingsfuncties van Starlette (inclusief **sessiecookies**).

> Gebouwd als een herbruikbare tool met componenten die makkelijk zijn te integreren in en met je systemen, datastores, relationele en NoSQL databases, etc.

Reviewer: ```suggestion
Gebouwd als een herbruikbare tool met componenten die makkelijk te integreren zijn in en met je systemen, datastores, relationele en NoSQL databanken, enz.
```

### Dependency Injection

> FastAPI bevat een uiterst eenvoudig, maar uiterst krachtig <abbr title='ook bekend als "components", "resources", "services", "providers"'><strong>Dependency Injection</strong></abbr> systeem.

Reviewer: Can we translate none of these "components", "resources", "services", "providers"?

* Zelfs dependencies kunnen dependencies hebben, waardoor een hiërarchie of **“graph” van dependencies** ontstaat.
* Allemaal **automatisch afgehandeld** door het framework.
* Alle dependencies kunnen data nodig hebben van request, de vereiste **padoperaties veranderen** en automatische documentatie verstrekken.
* **Automatische validatie** zelfs voor *padoperatie* parameters gedefinieerd in dependencies.

> * Ondersteuning voor complexe gebruikersauthenticatiesystemen, **databaseverbindingen**, etc.

Reviewer: ```suggestion
* Ondersteuning voor complexe gebruikersauthenticatiesystemen, **databankverbindingen**, etc.
```

> * **Geen compromis** met databases, frontends, etc. Maar eenvoudige integratie met ze allemaal.

Reviewer: ```suggestion
* **Geen compromisen** met databanken, gebruikersinterfaces, enz. Maar eenvoudige integratie met ze allemaal.
```

**FastAPI** is volledig verenigbaar met (en gebaseerd op) <a href="https://www.starlette.io/" class="external-link" target="_blank"><strong>Starlette</strong></a>.

`FastAPI` is eigenlijk een subklasse van `Starlette`. Dus als je Starlette al kent of gebruikt, zal de meeste functionaliteit op dezelfde manier werken.

Met **FastAPI** krijg je alle functies van **Starlette** (FastAPI is gewoon Starlette op steroïden):

* Zeer indrukwekkende prestaties. Het is <a href="https://github.com/encode/starlette#performance" class="external-link" target="_blank">een van de snelste Python frameworks, vergelijkbaar met **NodeJS** en **Go**</a>.
* **WebSocket** ondersteuning.
* Taken in de achtergrond tijdens het proces.
* Opstart- en afsluit events.
* Test client gebouwd op HTTPX.
* **CORS**, GZip, Statische bestanden, Streaming reacties.
* **Sessie en Cookie** ondersteuning.
* 100% van de code is getest.
* 100% type geannoteerde codebase.

## Pydantic functionaliteit

**FastAPI** is volledig verenigbaar met (en gebaseerd op) Pydantic. Dus alle extra <a href="https://docs.pydantic.dev/" class="external-link" target="_blank"><strong>Pydantic</strong></a> code die je nog hebt werkt ook.

> Inclusief externe pakketten die ook gebaseerd zijn op Pydantic, zoals <abbr title="Object-Relational Mapper">ORM</abbr>s, <abbr title="Object-Document Mapper">ODM</abbr>s voor databases.

Reviewer: ```suggestion
Inclusief externe pakketten die ook gebaseerd zijn op Pydantic, zoals <abbr title="Object-Relational Mapper">ORM</abbr>s, <abbr title="Object-Document Mapper">ODM</abbr>s voor databanken.
```

> Dit betekent ook dat je in veel gevallen het object dat je van een request krijgt **direct naar je database** kunt sturen, omdat alles automatisch wordt gevalideerd.

Reviewer: ```suggestion
Dit betekent ook dat je in veel gevallen het object dat je van een request krijgt **direct naar je databank** kunt sturen, omdat alles automatisch wordt gevalideerd.
```

> Hetzelfde geldt ook andersom, in veel gevallen kun je dus het object dat je krijgt van de database **direct** doorgeven aan de client**.

Reviewer: ```suggestion
Hetzelfde geldt ook andersom, in veel gevallen kun je dus het object dat je krijgt van de databank **direct doorgeven aan de client**.
```

Met **FastAPI** krijg je alle functionaliteit van **Pydantic** (omdat FastAPI is gebaseerd op Pydantic voor alle dataverwerking):

* **Geen brainfucks**:
    * Je hoeft geen nieuwe microtaal voor schemadefinities te leren.
    * Als je bekend bent Python types, weet je hoe je Pydantic moet gebruiken.
* Werkt goed samen met je **<abbr title=“Integrated Development Environment, vergelijkbaar met een code editor”>IDE</abbr>/<abbr title=“Een programma dat controleert op fouten in de code”>linter</abbr>/hersenen**:

>     * Doordat pydantic's data structuren enkel instanties zijn van classes, die je definieert, werken autocompletion, linting, mypy en je intuïtie allemaal goed met je gevalideerde data.

Reviewer: ```suggestion
    * Doordat pydantic's datastructuren enkel instanties zijn van klassen, die je definieert, werkt automatische aanvulling, linting, mypy en je intuïtie allemaal goed met je gevalideerde data.
```

> * Valideer **complexes structuren**:

Reviewer: ```suggestion
* Valideer **complexe structuren**:
```

    * Gebruik van hiërarchische Pydantic modellen, Python `typing`'s `List` en `Dict`, etc.
    * Met validators kunnen complexe dataschema's duidelijk en eenvoudig worden gedefinieerd, gecontroleerd en gedocumenteerd als JSON Schema.

>     * Je kunt diep geneste JSON** objecten laten valideren en annoteren.

Reviewer: ```suggestion
    * Je kunt diep **geneste JSON** objecten laten valideren en annoteren.
```
